uniffi_bindgen/src/interface/types.rs

```diff
@@ -128,243 +132,348 @@ impl From<&Type> for FFIType {
     }
 }
 
+impl Type {
+    /// Get the canonical, unique-within-this-component name for a type.
+    ///
+    /// When generating helper code for foreign language bindings, it's sometimes useful to be
+    /// able to name a particular type in order to e.g. call a helper function that is specific
+    /// to that type. We support this by defining a naming convention where each type gets a
+    /// unique canonical name, constructed recursively from the names of its component types (if any).
+    pub fn canonical_name(&self) -> String {
+        match self {
+            // Builtin primitive types, with plain old names.
+            Type::Int8 => "i8".into(),
+            Type::UInt8 => "u8".into(),
+            Type::Int16 => "i16".into(),
+            Type::UInt16 => "u16".into(),
+            Type::Int32 => "i32".into(),
+            Type::UInt32 => "u32".into(),
+            Type::Int64 => "i64".into(),
+            Type::UInt64 => "u64".into(),
+            Type::Float32 => "f32".into(),
+            Type::Float64 => "f64".into(),
+            Type::String => "string".into(),
+            Type::Boolean => "bool".into(),
+            // API defined types.
+            // Note that these all get unique names, and the parser ensures that the names do not
+            // conflict with a builtin type. We add a prefix to the name to guard against pathological
+            // cases like a record named `SequenceRecord` interfering with `sequence<Record>`
+            Type::Object(nm) => format!("Object{}", nm),
+            Type::Error(nm) => format!("Error{}", nm),
+            Type::Enum(nm) => format!("Enum{}", nm),
+            Type::Record(nm) => format!("Record{}", nm),
+            // Recursive types.
+            // These add a prefix to the name of the underlying type.
+            // The component API definition cannot give names to recursive types, so as long as the
+            // prefixes we add here are all unique amongst themselves, then we have no chance of
+            // acccidentally generating name collisions.
+            Type::Optional(t) => format!("Optional{}", t.canonical_name()),
+            Type::Sequence(t) => format!("Sequence{}", t.canonical_name()),
+            Type::Map(t) => format!("Map{}", t.canonical_name()),
```

Comment:
There's lots of places in the code where you're immediately calling `name.to_camel_case()`.

```
class_name_kt(&type_.canonical_name())?
```

Does heck respect the existing case? 

Just found: `lowerSequenceEnumEnumeration`, so yes it appears it does.